File: app/src/main/java/com/happs/ximand/ringcontrol/viewmodel/fragment/SelectDeviceViewModel.java
package com.happs.ximand.ringcontrol.viewmodel.fragment;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.databinding.ObservableBoolean;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.android.material.snackbar.Snackbar;
import com.happs.ximand.ringcontrol.FragmentNavigation;
import com.happs.ximand.ringcontrol.R;
import com.happs.ximand.ringcontrol.SingleLiveEvent;
import com.happs.ximand.ringcontrol.model.dao.BluetoothNDao;
import com.happs.ximand.ringcontrol.model.dao.SharedPreferencesDao;
import com.happs.ximand.ringcontrol.view.fragment.AllTimetablesFragment;
import com.happs.ximand.ringcontrol.viewmodel.dto.SnackbarDto;

import java.util.ArrayList;
import java.util.List;

import me.aflak.bluetooth.Bluetooth;

public class SelectDeviceViewModel extends BaseViewModel implements Bluetooth.DiscoveryCallback, Bluetooth.CommunicationCallback {

    private final MutableLiveData<List<BluetoothDevice>> devicesLiveData = new MutableLiveData<>();
    private final SingleLiveEvent<Void> startSettingsActivityLiveEvent = new SingleLiveEvent<>();
    private final ObservableBoolean searching = new ObservableBoolean();
    private final ObservableBoolean connecting = new ObservableBoolean();
    @Nullable
    private BluetoothDevice connectingDevice;

    public SelectDeviceViewModel() {
        devicesLiveData.setValue(new ArrayList<>());
        BluetoothNDao.getInstance().startScanning(this);
    }

    public LiveData<List<BluetoothDevice>> getDevicesLiveData() {
        return devicesLiveData;
    }

    public SingleLiveEvent<Void> getStartSettingsActivityLiveEvent() {
        return startSettingsActivityLiveEvent;
    }

    public ObservableBoolean getSearching() {
        return searching;
    }

    public ObservableBoolean getConnecting() {
        return connecting;
    }

    public void updateData() {
        this.devicesLiveData.setValue(getBondedDevicesList());
        getMakeSnackbarEvent().setValue(
                new SnackbarDto(R.string.data_updated, Snackbar.LENGTH_SHORT)
        );
    }

    private List<BluetoothDevice> getBondedDevicesList() {
        BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
        return new ArrayList<>(adapter.getBondedDevices());
    }

    public void startSettingsActivity() {
        startSettingsActivityLiveEvent.call();
    }

    public void notifyDeviceSelected(BluetoothDevice device) {
        connectingDevice = device;
        BluetoothNDao.getInstance().startConnecting(device, this);
        connectingDevice = device;
        connecting.set(true);
    }

    @Override
    public void onFinish() {
        searching.set(false);
    }

    @Override
    public void onDevice(BluetoothDevice device) {
        List<BluetoothDevice> devices = devicesLiveData.getValue();
        if (devices != null) {
            devices.add(device);
            devicesLiveData.postValue(devices);
        }
    }

    @Override
    public void onPair(BluetoothDevice device) {
        Log.d("...", "paired");
    }

    private void saveConnectedDeviceAddress() {
        if (connectingDevice != null) {
            SharedPreferencesDao.getInstance().updateTargetDeviceAddress(
                    connectingDevice.getAddress()
            );
        }
    }

    @Override
    public void onUnpair(BluetoothDevice device) {
        BluetoothNDao.getInstance().unpairDevice(device);
    }

    @Override
    public void onConnect(BluetoothDevice device) {
        connecting.set(false);
        if (device != null) {
            BluetoothNDao.getInstance().setPairedDevice(device);
            FragmentNavigation.getInstance()
                    .navigateTo(AllTimetablesFragment.newInstance());
            saveConnectedDeviceAddress();
        }
    }

    @Override
    public void onDisconnect(BluetoothDevice device, String message) {
        BluetoothNDao.getInstance().startConnecting(device, null);
        Log.d("...", "retrying");
    }

    @Override
    public void onMessage(String message) {

    }

    @Override
    public void onError(String message) {
        getMakeSnackbarEvent().setValue(
                new SnackbarDto(R.string.error_while_connecting, Snackbar.LENGTH_LONG)
        );
        searching.set(false);
    }

    @Override
    public void onConnectError(BluetoothDevice device, String message) {

    }
}

File: app/src/main/java/com/happs/ximand/ringcontrol/viewmodel/fragment/SelectDeviceViewModel.kt
package com.happs.ximand.ringcontrol.viewmodel.fragment

import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.bluetooth.BluetoothDevice
import android.content.Context
import androidx.databinding.ObservableBoolean
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.OnLifecycleEvent
import com.happs.ximand.ringcontrol.FragmentNavigation
import com.happs.ximand.ringcontrol.R
import com.happs.ximand.ringcontrol.model.`object`.exception.BaseException
import com.happs.ximand.ringcontrol.model.bl.BluetoothCommunicator
import com.happs.ximand.ringcontrol.model.bl.BluetoothScanner
import com.happs.ximand.ringcontrol.model.bl.callback.ConnectCallback
import com.happs.ximand.ringcontrol.model.bl.callback.SearchCallback
import com.happs.ximand.ringcontrol.model.dao.SharedPreferencesDao
import com.happs.ximand.ringcontrol.view.fragment.AllTimetablesFragment

class SelectDeviceViewModel : BaseViewModel(), SearchCallback, ConnectCallback {

    private val scanner = BluetoothScanner.instance
    private val communicator = BluetoothCommunicator.instance
    val devicesLiveData = MutableLiveData<MutableList<BluetoothDevice>>()

    var searchFinishedListener: (() -> Unit)? = null
    val connecting = ObservableBoolean()

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    fun onStart() {
        permissionRequest.value = ACCESS_FINE_LOCATION
    }

    fun updateData(applicationContext: Context, onFinish: () -> Unit) {
        if (scanner.isCurrentScanning()) {
            onFinish.invoke()
            return
        }
        preparePhoneForSearch()
        devicesLiveData.value = ArrayList()
        searchFinishedListener = onFinish
        scanner.searchCallback = this
        scanner.startDevicesSearching()
    }

    private fun preparePhoneForSearch() {
        if (!communicator.isBluetoothEnabled()) {
            communicator.enableBluetooth()
        }
    }

    fun connectToDevice(device: BluetoothDevice) {
        communicator.connectCallback = this
        communicator.connectToDevice(device)
    }

    override fun onFound(device: BluetoothDevice) {
        devicesLiveData.value?.add(device)
        devicesLiveData.value = devicesLiveData.value
    }

    override fun onFinish() {
        searchFinishedListener?.invoke()
    }

    override fun onConnected(device: BluetoothDevice) {
        connecting.set(true)
        FragmentNavigation.instance
                .navigateTo(AllTimetablesFragment.newInstance())
        SharedPreferencesDao.instance
                .updateTargetDeviceAddress(device.address)
    }

    override fun onException(e: BaseException) {
        makeExceptionSnackbarWithAction(R.string.unsuccessful_select, e)
        searchFinishedListener?.invoke()
    }

}